include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -67,36 +41,94 @@ struct exchangeValues<SharedMemoryAdaptor, float32_t>
+    mid <<= H - 1;
+
+    return mid | high | low;
+}
+
+template<uint32_t N, uint32_t H>
+enable_if_t<H <= N, uint32_t> bitShiftLeftHigher(uint32_t i)
+{
+    // Highest H bits are numbered N-1 through N - H
+    // N - 1 is then the highest bit, and N - 2 through N - H are the middle bits
+    // Lowest bits numbered from 0 through N - H - 1
+    uint32_t low = i & ((1 << (N - H)) - 1);
+    uint32_t mid = i & (~((1 << (N - H)) - 1) | ~(1 << (N - 1)));
+    uint32_t high = i & (1 << (N - 1));
+
+    mid <<= 1;
+    high >>= H - 1;
+
+    return mid | high | low;
+}
+
+// This function maps the index `idx` in the output array of a Forward FFT to the index `freqIdx` in the DFT such that `DFT[freqIdx] = output[idx]`
+// This is because Cooley-Tukey + subgroup operations end up spewing out the outputs in a weird order
+template<uint16_t ElementsPerInvocation, uint32_t WorkgroupSize>
+uint32_t getFrequencyIndex(uint32_t outputIdx)
+{
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t ELEMENTS_PER_INVOCATION_LOG_2 = uint32_t(mpl::log2<ElementsPerInvocation>::value);
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t FFT_SIZE_LOG_2 = ELEMENTS_PER_INVOCATION_LOG_2 + uint32_t(mpl::log2<WorkgroupSize>::value);
+
+    return bitShiftRightHigher<FFT_SIZE_LOG_2, FFT_SIZE_LOG_2 - ELEMENTS_PER_INVOCATION_LOG_2 + 1>(glsl::bitfieldReverse<uint32_t>(outputIdx) >> (32 - FFT_SIZE_LOG_2));
+}
+
+// This function maps the index `freqIdx` in the DFT to the index `idx` in the output array of a Forward FFT such that `DFT[freqIdx] = output[idx]`
+// It is essentially the inverse of `getFrequencyIndex`
+template<uint16_t ElementsPerInvocation, uint32_t WorkgroupSize>
+uint32_t getOutputIndex(uint32_t freqIdx)
+{
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t ELEMENTS_PER_INVOCATION_LOG_2 = uint32_t(mpl::log2<ElementsPerInvocation>::value);
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t FFT_SIZE_LOG_2 = ELEMENTS_PER_INVOCATION_LOG_2 + uint32_t(mpl::log2<WorkgroupSize>::value);
+
+    return glsl::bitfieldReverse<uint32_t>(bitShiftLeftHigher<FFT_SIZE_LOG_2, FFT_SIZE_LOG_2 - ELEMENTS_PER_INVOCATION_LOG_2 + 1>(freqIdx)) >> (32 - FFT_SIZE_LOG_2);
+}
+
+// Mirrors an index about the Nyquist frequency
+template<uint16_t ElementsPerInvocation, uint32_t WorkgroupSize>
+uint32_t mirror(uint32_t idx)
+{
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t FFT_SIZE = WorkgroupSize * uint32_t(ElementsPerInvocation);
+    return (FFT_SIZE - idx) & (FFT_SIZE - 1);
+}
+
+// When packing real FFTs a common operation is to get `DFT[T]` and `DFT[-T]` to unpack the result of a packed real FFT.
+// Given an index `idx` into the Nabla-ordered DFT such that `output[idx] = DFT[T]`, this function is such that `output[getNegativeIndex(idx)] = DFT[-T]`
+template<uint16_t ElementsPerInvocation, uint32_t WorkgroupSize>
+uint32_t getNegativeIndex(uint32_t idx)
+{
+    return getOutputIndex<ElementsPerInvocation, WorkgroupSize>(mirror<ElementsPerInvocation, WorkgroupSize>(getFrequencyIndex<ElementsPerInvocation, WorkgroupSize>(idx)));
+}
```

Comment:
I'd wrap all the functions that have `template<uint16_t ElementsPerInvocation, uint32_t WorkgroupSize>` into a single util struct and make them into static methods

Reply:
that way you can share the constants and masks like
```cpp
    NBL_CONSTEXPR_STATIC_INLINE uint32_t ELEMENTS_PER_INVOCATION_LOG_2 = uint32_t(mpl::log2<ElementsPerInvocation>::value);
    NBL_CONSTEXPR_STATIC_INLINE uint32_t FFT_SIZE_LOG_2 = ELEMENTS_PER_INVOCATION_LOG_2 + uint32_t(mpl::log2<WorkgroupSize>::value);
    ```
   as static constexpr member vars